lucene/core/src/java/org/apache/lucene/document/IndexSortDocValuesRangeQuery.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.Weight;
+
+/**
+ * A range query that takes advantage of the fact that the index is sorted to speed up execution.
+ * It assumes that the index is sorted on the same field as the query, and performs binary search
+ * on the field's doc values to find the documents at the lower and upper ends of the range.
+ *
+ * This optimized execution strategy can only be used if all of these conditions hold:
+ *   - The index is sorted, and its primary sort is on the same field as the query.
+ *   - The provided field has {@link SortedNumericDocValues}.
+ *   - The segments must have at most one field value per document (otherwise we cannot easily
+ *     determine the matching document IDs through a binary search).
+ */
+public final class IndexSortDocValuesRangeQuery extends Query {
+
+  private final String field;
+  private final long lowerValue;
+  private final long upperValue;
+
+  /**
+   * Creates a new {@link IndexSortDocValuesRangeQuery}.
+   *
+   * @param field The field name.
+   * @param lowerValue The lower end of the range (inclusive).
+   * @param upperValue The upper end of the range (inclusive).
+   */
+  public IndexSortDocValuesRangeQuery(String field,
+                                      long lowerValue,
+                                      long upperValue) {
+    this.field = field;
+    this.lowerValue = lowerValue;
+    this.upperValue = upperValue;
+  }
+
+  @Override
+  public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+    return new ConstantScoreWeight(this, boost) {
+      @Override
+      public Scorer scorer(LeafReaderContext context) throws IOException {
+        LeafReader reader = context.reader();
+
+        Sort indexSort = reader.getMetaData().getSort();
+        if (indexSort == null || indexSort.getSort().length == 0) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the index must be sorted.");
+        }
+
+        SortField sortField = indexSort.getSort()[0];
+        if (sortField.getField().equals(field) == false) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the index must be sorted " +
+              "on the same field as the query field.");
+        }
+
+        SortedNumericDocValues sortedDocValues = reader.getSortedNumericDocValues(field);
+        NumericDocValues docValues = DocValues.unwrapSingleton(sortedDocValues);
```

Comment:
I am curious to understand why we cannot support multiple values here. Is it because of duplicate elimination in binary searches?

Reply:
When a document has multiple values, the index sort must choose one of them to decide where it should be placed. So a binary search may miss a document where the value we’ve sorted on is outside the range, but it also contains a value that falls within the range.